controller/handler_edge_ctrl/common.go

```diff
@@ -0,0 +1,229 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"time"
+)
+
+type requestHandler interface {
+	getAppEnv() *env.AppEnv
+	getNetwork() *network.Network
+	getChannel() channel2.Channel
+	ContentType() int32
+	Label() string
+}
+
+type baseRequestHandler struct {
+	ch     channel2.Channel
+	appEnv *env.AppEnv
+}
+
+func (self *baseRequestHandler) getNetwork() *network.Network {
+	return self.appEnv.GetHostController().GetNetwork()
+}
+
+func (self *baseRequestHandler) getAppEnv() *env.AppEnv {
+	return self.appEnv
+}
+
+func (self *baseRequestHandler) getChannel() channel2.Channel {
+	return self.ch
+}
+
+func (self *baseRequestHandler) returnError(ctx requestContext, err error) {
+	responseMsg := channel2.NewMessage(int32(edge_ctrl_pb.ContentType_ErrorType), []byte(err.Error()))
+	responseMsg.ReplyTo(ctx.GetMessage())
+	if err = self.ch.Send(responseMsg); err != nil {
+		pfxlog.
+			ContextLogger(self.ch.Label()).
+			WithError(err).
+			WithField("token", ctx.GetSessionToken()).
+			WithField("operation", ctx.GetHandler().Label()).
+			Error("failed to send remove circuit response")
+	}
+}
+
+type requestContext interface {
+	GetHandler() requestHandler
+	GetSessionToken() string
+	GetMessage() *channel2.Message
+}
+
+type baseRequestContext struct {
+	handler      requestHandler
+	msg          *channel2.Message
+	err          error
+	sourceRouter *network.Router
+	session      *model.Session
+	service      *model.Service
+}
+
+func (self *baseRequestContext) GetMessage() *channel2.Message {
+	return self.msg
+}
+
+func (self *baseRequestContext) GetHandler() requestHandler {
+	return self.handler
+}
+
+func (self *baseRequestContext) loadRouter() bool {
+	routerId := self.handler.getChannel().Id().Token
+	self.sourceRouter, self.err = self.handler.getNetwork().GetRouter(routerId)
+	if self.err != nil {
+		logrus.
+			WithField("router", routerId).
+			WithField("operation", self.handler.Label()).
+			WithError(self.err).Errorf("could not find router closing channel")
+		_ = self.handler.getChannel().Close()
+		return false
+	}
+	return true
+}
+
+func (self *baseRequestContext) loadSession(token string) {
+	if self.err == nil {
+		self.session, self.err = self.handler.getAppEnv().Handlers.Session.ReadByToken(token)
+		if self.err != nil {
+			logrus.
+				WithField("token", token).
+				WithField("operation", self.handler.Label()).
+				WithError(self.err).Errorf("invalid session")
+		}
+	}
+}
+
+func (self *baseRequestContext) checkSessionType(sessionType string) {
+	if self.err == nil {
+		if self.session.Type != sessionType {
+			logrus.
+				WithField("token", self.session.Token).
+				WithField("operation", self.handler.Label()).
+				WithError(self.err).Errorf("wrong session type")
+			self.err = errors.New("invalid session")
+		}
+	}
+}
+
+func (self *baseRequestContext) checkSessionFingerprints(fingerprints []string) {
+	if self.err == nil {
+		now := time.Now()
+
+		var sessionCerts []string
```

Comment:
`sessionFingerprints` would be a better name. If you swap this loop with the loop below you cans just reuse the `prints` slice.

Reply:
Fixed